File: common-js/Message.js
const { randomUUID } = require('crypto');

class ResponseDestination {
    constructor(topic, uri) {
        this.topic = topic;
        this.uri = uri;
    }
}

class Message {
    constructor() {
        this.messageType = 'MESSAGE';
        this.sourceId = null;
        this.messageId = null;
        this.transactionId = null;
        this.uri = null;
        this.partition = null;
        this.responseDestination = null;
        this.data = null;
        this.timestamp = null;
        this.stream = null;
        this.streamState = null;
        this.streamIndex = null;
    }

    static newMessage(uri, sourceId, data) {
        const msg = new Message();
        msg.messageType = 'MESSAGE';
        msg.messageId = randomUUID();
        msg.transactionId = randomUUID();
        msg.uri = uri;
        msg.sourceId = sourceId;
        msg.data = data;
        msg.timestamp = Date.now();
        return msg;
    }

    static newRequest(uri, sourceId, transactionId, responseTopic, data) {
        const msg = new Message();
        msg.messageType = 'REQUEST';
        msg.messageId = randomUUID();
        msg.transactionId = transactionId || randomUUID();
        msg.uri = uri;
        msg.sourceId = sourceId;
        msg.data = data;
        msg.timestamp = Date.now();
        
        if (responseTopic) {
            msg.responseDestination = new ResponseDestination(responseTopic, uri);
        }
        return msg;
    }

    static newResponse(original, sourceId, data) {
        const msg = new Message();
        msg.messageType = 'RESPONSE';
        msg.messageId = original.messageId;
        msg.transactionId = original.transactionId;
        msg.uri = original.uri;
        msg.sourceId = sourceId;
        msg.data = data;
        msg.timestamp = Date.now();
        return msg;
    }
}

class Status {
    constructor(code = 'SUCCESS') {
        this.code = code;
        this.isSystemError = false;
        this.messageParams = null;
        this.source = null;
        this.params = null;
    }
}

class Response {
    constructor(data = null) {
        this.data = data;
        this.status = null;
    }

    hasError() {
        return this.status && this.status.code;
    }

    static dataResponse(data) {
        return new Response(data);
    }

    static systemErrorResponse(code) {
        const resp = new Response();
        resp.data = null;
        resp.status = new Status(code);
        resp.status.isSystemError = true;
        return resp;
    }
}

module.exports = { Message, ResponseDestination, Response, Status };

File: common-golang/message/message.go
package message

import (
	"time"

	"github.com/google/uuid"
)

type Type string

const (
	TypeMessage  Type = "MESSAGE"
	TypeRequest  Type = "REQUEST"
	TypeResponse Type = "RESPONSE"
)

type ResponseDestination struct {
	Topic string `json:"topic"`
	URI   string `json:"uri"`
}

type Message struct {
	MessageType         Type                 `json:"messageType,omitempty"`
	SourceID            string               `json:"sourceId,omitempty"`
	MessageID           string               `json:"messageId,omitempty"`
	TransactionID       string               `json:"transactionId,omitempty"`
	URI                 string               `json:"uri,omitempty"`
	Partition           *int                 `json:"partition,omitempty"`
	ResponseDestination *ResponseDestination `json:"responseDestination,omitempty"`
	Data                any                  `json:"data,omitempty"`
	Timestamp           *int64               `json:"timestamp,omitempty"`
	Stream              *bool                `json:"stream,omitempty"`
	StreamState         *string              `json:"streamState,omitempty"`
	TargetFile          *string              `json:"targetFile,omitempty"`
	Headers             map[string]string    `json:"headers,omitempty"`
}

func nowMillis() int64 {
	return time.Now().UnixMilli()
}

func NewMessage(uri, sourceID string, data any) Message {
	now := nowMillis()
	return Message{
		MessageType:   TypeMessage,
		MessageID:     uuid.New().String(),
		TransactionID: uuid.New().String(),
		URI:           uri,
		SourceID:      sourceID,
		Data:          data,
		Timestamp:     &now,
	}
}

func NewRequest(uri, sourceID, transactionID, responseTopic string, data any) Message {
	now := nowMillis()
	msg := Message{
		MessageType:   TypeRequest,
		MessageID:     uuid.New().String(),
		TransactionID: transactionID,
		URI:           uri,
		SourceID:      sourceID,
		Data:          data,
		Timestamp:     &now,
	}
	if responseTopic != "" {
		msg.ResponseDestination = &ResponseDestination{
			Topic: responseTopic,
			URI:   uri,
		}
	}
	return msg
}

func NewResponse(original Message, sourceID string, data any) Message {
	now := nowMillis()
	return Message{
		MessageType:   TypeResponse,
		MessageID:     original.MessageID,
		TransactionID: original.TransactionID,
		URI:           original.URI,
		SourceID:      sourceID,
		Data:          data,
		Timestamp:     &now,
	}
}
